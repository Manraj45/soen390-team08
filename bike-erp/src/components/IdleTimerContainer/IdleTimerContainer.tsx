import { Button, Modal, Typography } from '@material-ui/core';
import WarningIcon from '@material-ui/icons/Warning';
import React, { useRef, useState } from 'react';
import IdleTimer from 'react-idle-timer';
import { connect } from 'react-redux';
import { logout } from '../../redux/actions/AccountActions/accountAction';
import useStyles from './IdleTimerContainerStyle';

// Container for handling the inactivity period
const IdleTimerContainer = ({ account, logout }: any) => {
    // Reference to timer binded to react-idle-timer
    const idleTimerRef = useRef(null)

    // Reference to session countdown when prompt for inactivity
    const sessionTimeoutRef = useRef<any>(null)
    const sessionTimeIntervalRef = useRef<any>(null)

    // timeout period in ms,user will have a timeout period to extend their session
    const timeoutPeriod: number = 10000
    const [timeoutDisplay, settimeoutDisplay] = useState(timeoutPeriod)

    // Modal state
    const [modalOpen, setModalOpen] = useState(false)

    // Object for styling
    const classes = useStyles();

    // Function for when there is a period of inactivity
    const onIdle = () => {
        if (account.authenticated) {
            setModalOpen(true)
            sessionTimeoutRef.current = setTimeout(setInactive, timeoutPeriod);
            sessionTimeIntervalRef.current = setInterval(() => { settimeoutDisplay(timeoutDisplay => timeoutDisplay - 1000) }, 1000)
        }
    }

    // When user is inactive, call logout function from redux store
    const setInactive = () => {
        setModalOpen(false)
        clearTimeout(sessionTimeoutRef.current)
        clearInterval(sessionTimeIntervalRef.current)
        settimeoutDisplay(timeoutPeriod)
        logout()
    }

    // When Active, close modal and reset timeout
    const setActive = () => {
        setModalOpen(false)
        clearTimeout(sessionTimeoutRef.current)
        clearInterval(sessionTimeIntervalRef.current)
        settimeoutDisplay(timeoutPeriod)
    }

    // body representing the modal for inactivity
    const body = (
        <div className={classes.modal}>
            <WarningIcon fontSize="large" className={classes.icon} />
            <Typography className={classes.text}>
                You have been inactive.
                  <br />
                Your session will expire in {timeoutDisplay / 1000}
            </Typography>
            <Button className={classes.button} variant="contained" color="primary" onClick={setActive}>Extend</Button>
        </div >
    )

    return (
        <div id="idleTimerContainer">
            <IdleTimer ref={idleTimerRef} timeout={30000 /*after 30 sec -> call onIdle*/} onIdle={onIdle}></IdleTimer>
            <Modal open={modalOpen} aria-labelledby="simple-modal-title" aria-describedby="simple-modal-description">
                {body}
            </Modal>
        </div>
    )
}

const mapStateToProps = (state: any) => {
    return {
        account: state.account
    }
}

const mapDispatchToProps = (dispatch: any) => {
    return {
        logout: () => dispatch(logout())
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(IdleTimerContainer)