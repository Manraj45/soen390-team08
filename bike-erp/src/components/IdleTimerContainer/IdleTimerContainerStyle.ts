import { makeStyles } from "@material-ui/core";

const useStyles = makeStyles((theme) => ({
  modal: {
    background: "white",
    width: "30%",
    margin: "auto",
    border: "solid 1px white",
    marginTop: "20%",
    padding: "10px",
    outline: "none",
    minHeight: "140px",
  },
  text: {
    textAlign: "center",
  },
  button: {
    margin: "auto",
    alignItems: "center",
    display: "flex",
    marginTop: "10px",
  },
  icon: {
    margin: "auto",
    alignItems: "center",
    display: "flex",
  },
}));

export default useStyles;
