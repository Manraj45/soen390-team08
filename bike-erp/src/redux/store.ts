import { applyMiddleware, createStore } from "redux";
import rootReducer from "./rootReducer";
import thunk from "redux-thunk";

// Create store and apply thunk for API request
const store = createStore(rootReducer, applyMiddleware(thunk));

export default store;
