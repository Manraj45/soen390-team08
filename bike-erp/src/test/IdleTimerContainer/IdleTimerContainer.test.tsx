import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import renderer from "react-test-renderer";
import IdleTimerContainer from "../../components/IdleTimerContainer/IdleTimerContainer";
import store from "../../redux/store";

configure({ adapter: new Adapter() });

// Create mock registration page by injecting redux store
const setUp = () => {
  const testStore = store;
  const wrapper = shallow(<IdleTimerContainer store={testStore} />)
    .childAt(0)
    .dive();
  return wrapper;
};

describe("IdleTimerContainer Component test", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = setUp();
  });

  it("render IdleTimerContainer page without crashing", () => {
    const component = wrapper.find("#idleTimerContainer");
    expect(component.length).toBe(1);
  });

  it("match rendered component with snapshot", () => {
    const tree = renderer.create(
      <IdleTimerContainer store={store}></IdleTimerContainer>
    );
    expect(tree).toMatchSnapshot();
  });
});
