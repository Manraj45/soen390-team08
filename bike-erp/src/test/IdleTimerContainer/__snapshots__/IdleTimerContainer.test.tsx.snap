// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdleTimerContainer Component test match rendered component with snapshot 1`] = `
<div
  id="idleTimerContainer"
/>
`;
